From f258b5fc08aa3d54e83ebe874022ca2a0eb4f6ae Mon Sep 17 00:00:00 2001
From: Amlana Bhuyan <amlanab@ami.com>
Date: Tue, 10 Jun 2025 10:55:01 +0000
Subject: [PATCH] OT-5268: Session Information Display Enhancement

clientCutTextMsgHandlerHook is added for providing custom handling of
clientCutText messages.

Signed-off-by: Amlana Bhuyan <amlanab@ami.com>
---
 include/rfb/rfb.h            | 7 +++++++
 src/libvncserver/rfbserver.c | 6 ++++++
 2 files changed, 13 insertions(+)

diff --git a/include/rfb/rfb.h b/include/rfb/rfb.h
index bb264b17..741d8131 100644
--- a/include/rfb/rfb.h
+++ b/include/rfb/rfb.h
@@ -397,6 +397,7 @@ typedef struct sraRegion* sraRegionPtr;
 
 typedef void (*ClientGoneHookPtr)(struct _rfbClientRec* cl);
 typedef void (*ClientFramebufferUpdateRequestHookPtr)(struct _rfbClientRec* cl, rfbFramebufferUpdateRequestMsg* furMsg);
+typedef void (*ClientCutTextMsgHandlerHookPtr)(struct _rfbClientRec* cl, const char* text, uint32_t length);
 
 typedef struct _rfbFileTransferData {
   int fd;
@@ -689,6 +690,12 @@ typedef struct _rfbClientRec {
      */
     ClientFramebufferUpdateRequestHookPtr clientFramebufferUpdateRequestHook;
 
+    /**
+     * clientCutTextMsgHandlerHook is added for providing custom handling of
+     * clientCutText messages.
+     */
+    ClientCutTextMsgHandlerHookPtr clientCutTextMsgHandlerHook;
+
     rfbBool useExtDesktopSize;
     int requestedDesktopSizeChange;
     int lastDesktopSizeChangeError;
diff --git a/src/libvncserver/rfbserver.c b/src/libvncserver/rfbserver.c
index 5cb36a4a..06fe5cc7 100644
--- a/src/libvncserver/rfbserver.c
+++ b/src/libvncserver/rfbserver.c
@@ -2876,6 +2876,12 @@ rfbProcessClientNormalMessage(rfbClientPtr cl)
 	    return;
 	}
 	rfbStatRecordMessageRcvd(cl, msg.type, sz_rfbClientCutTextMsg+msg.cct.length, sz_rfbClientCutTextMsg+msg.cct.length);
+
+    if (cl->clientCutTextMsgHandlerHook)
+    {
+        cl->clientCutTextMsgHandlerHook(cl, str, msg.cct.length);
+    }
+
 #ifdef LIBVNCSERVER_HAVE_LIBZ
     if (isExtendedCutText) {
         if (msg.cct.length < 4) {
-- 
2.43.0

